src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -70915,5 +71101,222 @@ static void G<T>(T? x, object y)
                 Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "(default, x)").WithArguments("(T?, object? x)", "(T? x, object y)").WithLocation(12, 32));
             comp.VerifyTypes();
         }
+
+        [WorkItem(30140, "https://github.com/dotnet/roslyn/issues/30140")]
+        [Fact]
+        public void NullCoalescingAssignment_UseInExpression()
+        {
+            var source = @"
+class C
+{
+    void M(string? s1, string s2)
+    {
+        string s3 = (s1 ??= s2);
+        string? s4 = null, s5 = null;
+        string s6 = (s4 ??= s5); // Warn 1
+        s4.ToString(); // Warn 2
+    }
+}";
+
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (8,22): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         string s6 = (s4 ??= s5); // Warn 1
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "s4 ??= s5").WithLocation(8, 22),
+                // (9,9): warning CS8602: Possible dereference of a null reference.
+                //         s4.ToString(); // Warn 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s4").WithLocation(9, 9));
+        }
+
+        [WorkItem(30140, "https://github.com/dotnet/roslyn/issues/30140")]
+        [Fact]
+        public void NullCoalescingAssignment_AssignsState()
+        {
+            var source = @"
+class C
+{
+    object? F = null;
+    void M(C? c1, C c2, C c3)
+    {
+        c1 ??= c2;
+        c1.ToString();
+
+        if (c3.F == null) return;
+
+        c1 = null;
+        c1 ??= c3;
+        c1.F.ToString(); // Warn 1
+
+        c1 = null;
+        c1 ??= c3;
+        c1.ToString();
+        c1.F.ToString(); // Warn 2
+
+        if (c1.F == null) return; // Hidden 3
+        c1 ??= c2; // Hidden 4
+        c1.F.ToString(); // Warn 5
+    }
```

Comment:
>} [](start = 4, length = 1)

Consider testing:
```C#
if (c1.F == null) return;
(c1 ??= c3).F.ToString(); // ok
```

Reply:
Filed [https://github.com/dotnet/roslyn/issues/32501](https://github.com/dotnet/roslyn/issues/32501) to follow up on this.

---
In reply to: [247689418](https://github.com/dotnet/roslyn/pull/32225#discussion_r247689418) [](ancestors = 247689418)